src/mmfatl.h

```diff
@@ -0,0 +1,231 @@
+ * though.  Only pieces of text or unsigned integers can be embedded
+ * (%s or %u placeholder).
+ *
+ * For this kind of expansion you still need a buffer where the final message
+ * is constructed.  In our context, this buffer is pre-allocated, and fixed in
+ * size, truncation of overflowing text enforced.
+ *
+ * Documentation
+ * =============
+ *
+ * Most comments are written in <a href="https://doxygen.nl/index.html">
+ * doxygen</a> (Qt variant) style.  If you have doxygen installed on your
+ * computer, you may generate HTML documentation out of them with its root
+ * page placed in build/html/index.html by running build.sh with the -d option.
+ *
+ * Regression tests
+ * ================
+ *
+ * If the macro **REGRESSION_TEST** is defined (option -t of build.sh) or
+ * **TEST_MMFATL** is defined, then regression tests are implemented.  Invoke
+ * in addition option -c on build.sh, should you switch between with/out
+ * testing, but no intermediate source file change.
+ *
+ * In order to run implemented regression tests properly we suggest to add a
+ * \code{.c}
+ * mmfatl_test();
+ * \endcode
+ * line to main() close to its begin and **before** any function declared in
+ * this header file is called.
+ *
+ * If tests are disabled this line evaluates to nothing.  In addition, the
+ * compiler skips any test code, so the artifact size will not grow.  All in
+ * all, a disabled test suite does not come with a linking or runtime penalty.
+ *
+ * If enabled, running tests document their progress to stdout.  Testing stops
+ * on the first regression found with a diagnostic message further detailing on
+ * the context of the failure.  The tests, if called as described above, do not
+ * interfere with Metamath program behaviour, even not when detecting a
+ * regression, in any way other than generating extra output on program start.
+ *
+ * We recommend running the tests each time you modify mmfatl.h or mmfatl.c
+ * to ensure it still executes as desired.
+ */
+
+#include <stddef.h>
+
+/* REGRESSION_TEST overrides the setting of TEST_MMFATL */
+#ifdef REGRESSION_TEST
+#   define TEST_MMFATL
+#endif
+
+/* uncomment one of the following to enforce disabling/enabling of regression
+ * tests unconditionally
+ */
+// #undef TEST_MMFATL
+// #define TEST_MMFATL
+
+#ifdef TEST_MMFATL
+    /* regression tests are implemented and called through this function */
+    extern void mmfatl_test(void);
+#else
+    /* an empty macro deletes any call to the regression test suite at
+     * compile time */
+#   define mmfatl_test(x)
+#endif
+
+
+/*--------------------   Allocation Of An Error Buffer   --------------------*/
+/*!
+ * Basic parameters controlling pre-allocation of basic routines and data.
+ * Pre-allocation helps to construct a diagnosis message, even in a
+ * corrupt environment.
+ *
+ * A pre-allocated buffer is provided to allow embedding data in a diagnosis
+ * message.
+ */
+struct FatalErrorBufferDescriptor {
```

Comment:
IIRC I don't think metamath.exe uses uppercase for type names, I think they are snake case.

Reply:
Ok, this was my choice, but I do not insist on it.

Reply:
Then you should do it in a separate PR and you should do it uniformly to everything.

Reply:
I've looked into mmdata.h, and it is more than 90% camel case and less than 10% snake case.  Here are examples:
processUndoStack, free_pntrString.  Is there a howto somewhere giving a style guide?

Reply:
Those are functions, not types.

If the style is inconsistent, then I'm happy to change it, but like I said that should be done in a PR which is only about that and is implemented across the board, not just in a few files.

Reply:
Whatever this takes it is now deferred to a later contribution.